### PR TITLE
Allow image maxWidth and maxHeight

### PR DESCRIPTION
**Pull request by eexit**
Hi,

Thank you for fancybox, it's great and especially nice to dig into the code :)
What I would like to add is the ability to allow a specific max-size when images are opened.

Here's an example where I set the max-width to my container width:

screen shot 2017-07-28 at 1 36 32 am

This is very useful because I use [RetinaJS](http://imulus.github.io/retinajs/) and my images are up-sampled to a certain size factor that fancybox blasted out by loading the image to cover the max viewport area as possible.

I added two options in the `image` section of the fancbybox options that allows to do that.

```js
image : {

    // Wait for images to load before displaying
    // Requires predefined image dimensions
    // If 'auto' - will zoom in thumbnail if 'width' and 'height' attributes are found
    preload : "auto",

    // Maximum image width. Allows image downscalling
    // Image height is automatically computed from this value
    maxWidth : 0,

    // Maximum image height. Allows image downscalling
    // Image width is automatically computed from this value
    maxHeight : 0,

},
```

Demo here (on side-by-side images): http://joris.berthelot.photography/destination-moon/

Review and feedback is very welcomed :)

Thanks!

### CODE REVIEW

**Comment by eexit** [comment]
Hello @fancyapps? Any feedback on this?

Thanks!

**Comment by tobek** [comment]
This is great! Would also be great to be able to specify max width/height for an individual image e.g. with a `data-max-width` attribute. This is pretty easy, since data attributes are automatically consumed, so when you call `self.resolveImageSlideSize` you have access e.g. to `slide.opts.maxWidth` pulled from the data attribute.

I've implemented this for myself - not sure if I can add a commit to this PR or if I should make a new PR, also not sure if PRs on this project are getting merged!

**Comment by fancyapps** [comment]
Hi,

You can (after updating to the latest version) re-use `data-width` and `data-height` attributes for the same purpose.

**Comment by eexit** [comment]
@fancyapps Do you have the doc for that, I can't find it.

Thanks!

**Comment by fancyapps** [comment]
It is not yet in the docs, sorry 😄 

But, as a nice side-effect, it can be used to set dimension for SVG files - https://codepen.io/anon/pen/QOVjdx?editors=1100

**Comment by eexit** [comment]
@fancyapps It doesn't seem to work properly if you omit either `data-width` or `data-height`.

**Comment by tobek** [comment]
Hey, appreciate the update! However, unlike this PR, the update doesn't preserve image aspect ratio, so you have to set both and work out the ratio, which isn't ideal. I guess that makes `width`/`height` and accurate attribute name, but `max-width`/`max-height` is probably what a lot of people are looking for.

**Comment by fancyapps** [comment]
So, you are looking for something like this - https://codepen.io/anon/pen/zPJXaN?editors=1000 ?

**Comment by tobek** [comment]
Cool that looks great. What do you think about that being the default behavior instead of having to include e.g. `data-width="0"`?

**Comment by eexit** [comment]
I believe these attributes are filled by fancybox no? In that case, it may be tricky to differentiate the real image dimensions than the one we would like to constraint, could be also misleading to users.
That's the reason why, using new attributes could fit this usage: `data-max-*` or `data-slide-*`.

**Comment by fancyapps** [comment]
@tobek Yeah, ok. 

btw, maybe you could just adjust stage dimensions, e.g.:

```CSS
.fancybox-stage {
    width: 480px;
    margin: auto;
}
```

then it would look like it covers your container.

**Comment by tobek** [comment]
The thing about adjusting the stage is that (in my case at least) I need different sizes for different images in the same gallery. For example I have a work portfolio where web design mocks can be shown full size, but logos should be smaller because they look silly filling the page. If I simply use smaller versions for some images, it looks bad on retina screens because the image gets blown up to full resolution. Even without retina though, ideally the user shouldn't have to resize images in order to use them here.

(There are probably other ways to handle retina issues in general, but that would be a separate feature.)

We appreciate your work and the back-and-forth on this!

**Comment by fancyapps** [comment]
So I actually updated the code so it matches the logic of your snippet, you can see that in previous codepen link. 

But, sorry, I still don't understand, what do you mean by "logos should be smaller because they look silly filling the page" and "If I simply use smaller versions for some images, it looks bad on retina screens because the image gets blown up to full resolution". Why would logos fill the page and why would images blown up? Could you maybe show some demo?  

I guess you are looking for `srcset` /`sizes` feature (see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img) and fancyBox currently has only basic support.  

Also, it seems to me that you might need something like `openTo: 'cover'` as described here - https://github.com/fancyapps/fancybox/issues/1740

**Comment by eexit** [comment]
What @tobek means is that in order to display retina compliant images, we need to force downsize the image dimension so the images looks nice on retina displays.
If an image is 400px width, the retina version must be 800px width in order to display nicely BUT in a 400px width image tag.

The feature we are asking is the ability to force the slide dimensions to a configurable size so whatever the image dimensions are, it still shows up in a forced dimensions slide/keeping the ratio, of course.

Also, it would be really nice to be able to have this feature as a global param. For instance, on my portfolio, fancy images will never be displayed wider than 800px. So instead of parsing all image tags to add the data attribute, it'd be nice to set the global param.

Could be even better: global param < local param :)

Does this make sense?
Thank you very much!

**Comment by fancyapps** [comment]
In that case, I think you actually need:

1) option like `retinaRatio: 2` that would automatically down scale images 2x;

2) and/or some way (like some callback) to customize source link depending on some conditions, for example, you could change image source link to @2x version depending on `window.devicePixelRatio`. Since not all images probably would have @2x version , you could, for example, add `data-retina-src` attribute to links to mark that needs to be adjusted.

**Comment by tobek** [comment]
Option 2 is great actually - with a manual `retinaRatio` flag the user would have to detect `window.devicePixelRatio` themselves and then set that flag. `data-retina-src` to mark the images that have retina sources and then automatically triggering a retina ratio with a `@2x` version of the file name would be great, assuming good browser compatibility for detecting the necessary info. I'm not sure if you have a retina device, but some images do truly look bad without handling this. Also FWIW, some devices now have 3x device pixel ratios now too.

So yeah, there were two reasons I personally wanted this sizing feature:

1. To handle retina images because there's currently no way to do it
2. Because aesthetically not all images need to be displayed at the same size (the "logos should be smaller because they look silly filling the page" issue). I have a portfolio of web dev and design work. Web dev work [like this](https://user-images.githubusercontent.com/3081073/33383477-6aa40c30-d4f1-11e7-8155-6e0c297eb683.jpg) can be displayed full screen. A logo [like this](https://user-images.githubusercontent.com/3081073/33383504-7eb6c046-d4f1-11e7-87d5-6e8a906ebf9e.jpg) looks silly full screen, so it should be downsized [like this](https://user-images.githubusercontent.com/3081073/33383523-90c7d108-d4f1-11e7-88da-6ae88878d4c4.jpg). Because of the retina issue I couldn't simply link to a smaller image, because it would get blown up. But even with the retina issue solved, it would be nice to not have to resize source images and instead just specify the size in code.

I agree with @eexit that a global setting would be useful for some - I see in his site a smaller size matches the layout. In that use case I think styling `.fancybox-stage` would do the trick, and as long as that's documented that's cool.

Bottom line: your updates that support per-image sizing while maintaining aspect ratios - that can handle all these situations, so we're good I think. But improved/easier retina support is definitely something to consider doing.

**Comment by fancyapps** [comment]
OK, so I was thinking a lot about this and decided that the simplest solution would be to rearrange the code a bit so that user can use `beforeLoad` callback to change image source and use `afterLoad` to change image dimensions, if needed. This provides more options for customization. For example, image source or dimensions can be adjusted depending on image size or some attribute (for example, some attribute might indicate that image has 2x or 3x version) 

Example or scaling down the image for retina displays:

```JS
$('[data-fancybox="images"]').fancybox({
  afterLoad : function(instance, current) {
    var pixelRatio = window.devicePixelRatio || 1;

    if ( pixelRatio > 1.5 ) {
      current.width  = current.width  / pixelRatio;
      current.height = current.height / pixelRatio;
    }
  }
});
```

Demo -  https://codepen.io/anon/pen/mqQLVm?editors=1010

Another fun demo - display larger unsplash images for retina displays while keeping current size:

```JS
$('[data-fancybox="images"]').fancybox({
  beforeLoad : function(instance, current) {
    var pixelRatio = window.devicePixelRatio || 1,
        rez,
        width,
        height;

    if ( pixelRatio > 1 && ( rez = current.src.match( /\/(\d+)x(\d+)*/ ) ) ) {
      width  = parseInt( rez[1], 10 );
      height = parseInt( rez[2], 10 );

      current.opts.width  = width;
      current.opts.height = height;

      current.src = current.src.replace( rez[0], '/' + ( width * pixelRatio ) + 'x' + ( height * pixelRatio ) );
    }

  }
});
```

Demo - https://codepen.io/anon/pen/yPQjbw?editors=1010

**Comment by tobek** [comment]
This is cool and very flexible!

I'd vote for still having `data-width`/`data-height` in there as an easy way to accomplish one common task. And a lot of users might be using/writing basically the same `afterLoad` function for retina purposes, so having an example in the docs that describes it, or data-attributes that trigger some general retina logic, would be useful for users who are less savvy.

Otherwise, very :100:

**Comment by eexit** [comment]
@fancyapps Yeah, why not. But don't need to reinvent the wheel here. RetinaJS works like a charm so just be able to stick with forced dimension and ratio is good for me :)

If you want, you can even try a RetinaJS integration or develop a bridge. I think this would add a lot of value to Fancybox 😉

**Comment by tobek** [comment]
Never actually used RetinaJS but that could surely help. Looks like you still have to manage constraining the dimensions yourself though (or at least specify in CSS, which could maybe mess with Fancybox logic), so some control of dimensions in Fancybox, whether in an `afterLoad` function or via data attributes, would be necessary.
